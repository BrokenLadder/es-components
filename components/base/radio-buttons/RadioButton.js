import React from 'react';
import classNames from 'classnames';
import { noop } from 'lodash';
import '../../../styles/inputs.less';
import './radio-list.less';

function RadioButton({
  id,
  disabled,
  shouldDisplayValidation,
  checked,
  inline,
  labelText,
  name,
  onChange,
  value,
  ...radioProps
}) {
  const radioLabelClasses = classNames('label-style', 'radio-label', {
    'disabled-label': disabled
  });

  const radioDisplayClass = classNames({
    'errored-radio': !checked && !disabled && shouldDisplayValidation,
    'empty-radio': !checked && !disabled && !shouldDisplayValidation,
    'disabled-radio': !checked && disabled && !shouldDisplayValidation,
    'filled-radio': checked && !disabled && !shouldDisplayValidation,
    'disabled-filled-radio': disabled && checked && !shouldDisplayValidation
  });

  const radioWrapperClasses = classNames({
    'error': shouldDisplayValidation,
    'inline-radio-list': inline
  });

  return (
    <div className={radioWrapperClasses}>
      <label className={radioLabelClasses} htmlFor={id}>
        <input
          type="radio"
          name={name}
          className="radio-input"
          onChange={onChange}
          value={value}
          disabled={disabled}
          {...radioProps}
        />
        <span className="radio-text">{labelText}</span>
        <span className={radioDisplayClass} />
      </label>
    </div>
  );
}

RadioButton.propTypes = {
  labelText: React.PropTypes.string.isRequired,
  name: React.PropTypes.string.isRequired,
  id: React.PropTypes.string,
  disabled: React.PropTypes.bool,
  inline: React.PropTypes.bool,
  onChange: React.PropTypes.func,
  value: React.PropTypes.any,
  /**
   * Should the radio button be displayed as invalid
   */
  shouldDisplayValidation: React.PropTypes.bool
};

RadioButton.defaultProps = {
  disabled: false,
  inline: true,
  shouldDisplayValidation: false,
  onChange: noop
};

export default RadioButton;
