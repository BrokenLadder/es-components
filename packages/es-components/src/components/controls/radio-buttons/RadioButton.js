import React from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

import Label from '../Label';
import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 50%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: relative;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? props.theme.colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'flex')};
  font-size: ${props => props.theme.sizes.baseFontSize};
  font-weight: normal;
  margin-bottom: 18px;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .es-radio__fill:before {
    ${props => radioFill(props.hoverFillColor)};
  }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .es-radio__fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
  }
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 50%;
  box-sizing: border-box;
  height: 25px;
  margin-right: 5px;
  width: 25px;

  &:before {
    ${props => radioFill(props.fill)};
  }
`;

export function RadioButton({
  optionText,
  name,
  isChecked = false,
  id,
  isDisabled = false,
  inline = true,
  hasError = false,
  theme,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    isChecked,
    isDisabled,
    hasError,
    theme.colors
  );
  const radioDisplayFill = isChecked ? fill : theme.colors.white;

  const labelProps = {
    inline,
    disabled: isDisabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: hasError
  };

  return (
    <RadioLabel className="es-radio" {...labelProps}>
      <RadioInput
        type="radio"
        name={name}
        id={id}
        disabled={isDisabled}
        checked={isChecked}
        {...radioProps}
      />
      <RadioDisplay
        className="es-radio__fill"
        borderColor={fill}
        fill={radioDisplayFill}
      />
      {optionText}
    </RadioLabel>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  isChecked: PropTypes.bool,
  id: PropTypes.string,
  isDisabled: PropTypes.bool,
  inline: PropTypes.bool,
  hasError: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object.isRequired
};

RadioButton.defaultProps = {
  isChecked: false,
  id: undefined,
  isDisabled: false,
  inline: false,
  hasError: false
};

export default withTheme(RadioButton);
