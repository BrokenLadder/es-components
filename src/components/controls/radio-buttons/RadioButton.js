import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider } from 'styled-components';
import { noop } from 'lodash';

import defaultTheme from '../../theme/defaultTheme';
import Label from '../Label';
import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 100%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: absolute;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? props.theme.colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'inherit')};
  font-size: ${props => props.theme.sizes.baseFontSize}px;
  font-weight: normal;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  min-height: 25px;
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .radio-fill:before {
    ${props => radioFill(props.hoverFillColor)};
  }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .radio-fill {
    box-shadow: 0 0 3px 3px #83bffc;
  }
`;

const RadioText = styled.span`
  align-self: center;
  margin-left: 30px;
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 100%;
  box-sizing: border-box;
  display: block;
  height: 25px;
  position: absolute;
  width: 25px;

  &:before {
    ${props => radioFill(props.fill)};
  }
`;

function RadioButton({
  optionText,
  name,
  checked = false,
  id,
  isDisabled = false,
  inline = true,
  onClick = noop,
  value,
  hasError = false,
  theme,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    theme.colors,
    checked,
    isDisabled,
    hasError
  );
  const radioDisplayFill = checked ? fill : theme.colors.white;

  const labelProps = {
    checked,
    inline,
    disabled: isDisabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: hasError
  };

  return (
    <ThemeProvider theme={theme}>
      <RadioLabel {...labelProps}>
        <RadioInput
          type="radio"
          name={name}
          id={id}
          onClick={onClick}
          value={value}
          disabled={isDisabled}
          {...radioProps}
        />
        <RadioDisplay
          className="radio-fill"
          borderColor={fill}
          fill={radioDisplayFill}
        />
        <RadioText>{optionText}</RadioText>
      </RadioLabel>
    </ThemeProvider>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  id: PropTypes.string,
  isDisabled: PropTypes.bool,
  inline: PropTypes.bool,
  onClick: PropTypes.func,
  value: PropTypes.any,
  hasError: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

RadioButton.defaultProps = {
  theme: defaultTheme
};

export default RadioButton;
