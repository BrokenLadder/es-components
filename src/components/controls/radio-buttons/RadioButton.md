### Default
```
<RadioButton
  name="cheeses"
  labelText="Grueyere"
/>
```

### Disabled
```
<RadioButton
  name="cheese"
  labelText="Gouda"
  disabled
/>
```

### Checked
```
<RadioButton
  name="cheese"
  labelText="Parmesan"
  checked
/>
```

### Checked and Disabled
```
<RadioButton
  name="cheese"
  labelText="American"
  checked
  disabled
/>
```

### Errored
```
<RadioButton
  name="cheese"
  labelText="Bleu"
  shouldDisplayValidation
/>
```