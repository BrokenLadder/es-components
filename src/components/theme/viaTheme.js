const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
