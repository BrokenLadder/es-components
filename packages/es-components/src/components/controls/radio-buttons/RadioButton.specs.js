/* eslint-env jest */
/* eslint-disable max-len */

import React from 'react';
import { renderWithTheme } from 'styled-enzyme';
import viaTheme from 'es-components-via-theme';

import getRadioFillVariables from './radio-fill-variables';
import { RadioButton } from './RadioButton';

describe('RadioButton component', () => {
  it('renders as expected', () => {
    const tree = renderWithTheme(
      <RadioButton id="test" optionText="test" name="rad" theme={viaTheme} />
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('getRadioFillVariables', () => {
  it('returns { fill: gray5, hover: white } when unchecked, disabled, and valid', () => {
    const checked = false;
    const disabled = true;
    const validationState = 'default';

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray5);
    expect(hover).toBe(viaTheme.colors.white);
  });

  it('returns { fill: info } when checked, not disabled, and valid', () => {
    const checked = true;
    const disabled = false;
    const validationState = 'default';

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.info);
  });

  it('returns { fill: gray5 } when checked, disabled, and valid', () => {
    const checked = true;
    const disabled = true;
    const validationState = 'default';

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray5);
  });

  it('returns { fill: danger, hover: gray8 } when unchecked, not disabled, and invalid', () => {
    const checked = false;
    const disabled = false;
    const validationState = 'danger';

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.danger);
    expect(hover).toBe(viaTheme.colors.gray8);
  });

  it('returns { fill: info } when checked, not disabled, and invalid', () => {
    const checked = true;
    const disabled = false;
    const validationState = 'danger';

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.info);
  });

  it('returns { fill: gray8, hover: gray8 } when unchecked, not disabled, and valid', () => {
    const checked = false;
    const disabled = false;
    const validationState = 'default';

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      validationState,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray8);
    expect(hover).toBe(viaTheme.colors.gray8);
  });
});
