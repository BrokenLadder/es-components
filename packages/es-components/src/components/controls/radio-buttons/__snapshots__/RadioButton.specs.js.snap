// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton component renders as expected 1`] = `
<label
  checked={false}
  className="sc-bwzfXH ctIRCP sc-bdVaJa kHjvzh"
  disabled={false}
  htmlFor="test"
>
  <input
    className="sc-htpNat jeuMMj"
    disabled={false}
    id="test"
    name="rad"
    onClick={[Function]}
    type="radio"
    value={undefined}
  />
  <span
    className="radio-fill sc-ifAKCX cjMraT"
    fill="#fff"
  />
  <span
    aria-hidden={undefined}
    className="sc-bxivhb bRNnnP"
  >
    test
  </span>
</label>
`;
